src/System.Security.Permissions/tests/PrincipalPermissionTests.cs

```diff
@@ -0,0 +1,329 @@
+            PrincipalPermission p2 = new PrincipalPermission("user", "role");
+            PrincipalPermission p3 = (PrincipalPermission)p1.Union(p2);
+            Assert.True(p3.IsUnrestricted());
+            p3 = (PrincipalPermission)p2.Union(p1);
+            Assert.True(p3.IsUnrestricted());
+        }
+
+        [Fact]
+        public void Union()
+        {
+            PrincipalPermission p1 = new PrincipalPermission("user A", "role A");
+            PrincipalPermission p2 = new PrincipalPermission("user B", "role B", false);
+            PrincipalPermission p3 = (PrincipalPermission)p1.Union(p2);
+            Assert.True(p3.ToString().IndexOf("user A") >= 0);
+            Assert.True(p3.ToString().IndexOf("user B") >= 0);
+        }
+
+        [Fact]
+        public void UnionWithBadPermission()
+        {
+            PrincipalPermission p1 = new PrincipalPermission("user", null);
+            EnvironmentPermission ep2 = new EnvironmentPermission(PermissionState.Unrestricted);
+            Assert.Throws<ArgumentException>(() => p1.Union(ep2));
+        }
+
+        [Fact]
+        public void IntersectWithNull()
+        {
+            PrincipalPermission p1 = new PrincipalPermission("user", "role");
+            PrincipalPermission p2 = null;
+            PrincipalPermission p3 = (PrincipalPermission)p1.Intersect(p2);
+            Assert.Null(p3);
+        }
+
+        [Fact]
+        public void IntersectWithUnrestricted()
+        {
+            PrincipalPermission p1 = new PrincipalPermission(PermissionState.Unrestricted);
+            PrincipalPermission p2 = new PrincipalPermission("user", "role");
+            PrincipalPermission p3 = (PrincipalPermission)p1.Intersect(p2);
+            Assert.True(!p3.IsUnrestricted());
+            Assert.Equal(p2.ToXml().ToString(), p3.ToXml().ToString());
+            p3 = (PrincipalPermission)p2.Intersect(p1);
+            Assert.True(!p3.IsUnrestricted());
+            Assert.Equal(p2.ToXml().ToString(), p3.ToXml().ToString());
+        }
+
+        [Fact]
+        public void Intersect_NoIntersection()
+        {
+            // no intersection
+            PrincipalPermission p1 = new PrincipalPermission("user A", "role 1");
+            PrincipalPermission p2 = new PrincipalPermission("user B", "role 2");
+            PrincipalPermission p3 = (PrincipalPermission)p1.Intersect(p2);
+            Assert.Null(p3);
+        }
+
+        [Fact]
+        public void Intersect_IntersectionInRole()
+        {
+            PrincipalPermission p1 = new PrincipalPermission("user A", "role 1");
+            PrincipalPermission p2 = new PrincipalPermission("user C", "role 1");
+            PrincipalPermission p3 = (PrincipalPermission)p2.Intersect(p1);
+            Assert.NotNull(p3);
+            Assert.True(p3.ToString().IndexOf("user A") < 0);
+            Assert.True(p3.ToString().IndexOf("user C") < 0);
+            Assert.True(p3.ToString().IndexOf("role 1") >= 0);
```

Comment:
Save the p3.ToString() value to a variable.  Then use Assert.Contains and Assert.DoesNotContain